File: result/parser.go
package result

import "fmt"

const (
	responseErr = "unexpected google images response format\n%v"
)

// Parser parsers query responses, every query implementation must be aware of its expected input and enforce it
// In case of invalid input an error should be returned
type Parser interface {
	Parse(interface{}) (interface{}, error)
}

// GoogleImagesResultParser is the specific parser to parse Google Images results
type GoogleImagesResultParser struct {
}

// Parse takes the first item content and extracts the link field from it
func (g GoogleImagesResultParser) Parse(raw interface{}) (interface{}, error) {
	r, ok := raw.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf(responseErr, r)
	}

	items, ok := r["items"].([]interface{})
	if !ok {
		return nil, fmt.Errorf(responseErr, r)
	}

	if len(items) < 1 {
		return nil, fmt.Errorf(responseErr, r)
	}

	item, ok := items[0].(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf(responseErr, r)
	}

	link, ok := item["link"].(string)
	if !ok {
		return nil, fmt.Errorf(responseErr, r)
	}

	return link, nil
}

File: typescript/src/result/parser.ts
export interface Parser {
  parse(data: any): Promise<any>;
}

const RESPONSE_ERR = 'unexpected google images response format';

export class GoogleImagesResultParser implements Parser {
  async parse(raw: any): Promise<string> {
    if (!raw || typeof raw !== 'object') {
      throw new Error(`${RESPONSE_ERR}: ${JSON.stringify(raw)}`);
    }

    const items = raw.items;
    if (!Array.isArray(items)) {
      throw new Error(`${RESPONSE_ERR}: items not found or not array`);
    }

    if (items.length < 1) {
      throw new Error(`${RESPONSE_ERR}: no items returned`);
    }

    const item = items[0];
    if (!item || typeof item !== 'object') {
      throw new Error(`${RESPONSE_ERR}: invalid item format`);
    }

    const link = item.link;
    if (typeof link !== 'string') {
      throw new Error(`${RESPONSE_ERR}: link not found or not string`);
    }

    return link;
  }
}